ndb/src/google/cloud/ndb/model.py

```diff
@@ -2947,28 +2967,405 @@ def __init__(self, *args, **kwargs):
+
+        from google.cloud import ndb
+
+
+        class MyModel(ndb.Model):
+            value = ndb.FloatProperty()
+            description = ndb.StringProperty()
+
+    .. doctest:: model-key
+
+        >>> entity1 = MyModel(id=11)
+        >>> entity1.key
+        Key('MyModel', 11)
+        >>> entity2 = MyModel(parent=entity1.key)
+        >>> entity2.key
+        Key('MyModel', 11, 'MyModel', None)
+        >>> entity3 = MyModel(key=ndb.Key(MyModel, "e-three"))
+        >>> entity3.key
+        Key('MyModel', 'e-three')
+
+    However, a user-defined property can be defined on the model with the
+    same name as one of those keyword arguments. In this case, the user-defined
+    property "wins":
+
+    .. doctest:: model-keyword-id-collision
+
+        >>> class IDCollide(ndb.Model):
+        ...     id = ndb.FloatProperty()
+        ...
+        >>> entity = IDCollide(id=17)
+        >>> entity
+        IDCollide(id=17.0)
+        >>> entity.key is None
+        True
+
+    In such cases of argument "collision", an underscore can be used as a
+    keyword argument prefix:
+
+    .. doctest:: model-keyword-id-collision
+
+        >>> entity = IDCollide(id=17, _id=2009)
+        >>> entity
+        IDCollide(key=Key('IDCollide', 2009), id=17.0)
+
+    For the **very** special case of a property named ``key``, the ``key``
+    attribute will no longer be the entity's key but instead will be the
+    property value. Instead, the entity's key is accessible via ``_key``:
+
+    .. doctest:: model-keyword-key-collision
+
+        >>> class KeyCollide(ndb.Model):
+        ...     key = ndb.StringProperty()
+        ...
+        >>> entity1 = KeyCollide(key="Take fork in road", id=987)
+        >>> entity1
+        KeyCollide(_key=Key('KeyCollide', 987), key='Take fork in road')
+        >>> entity1.key
+        'Take fork in road'
+        >>> entity1._key
+        Key('KeyCollide', 987)
+        >>>
+        >>> entity2 = KeyCollide(key="Go slow", _key=ndb.Key(KeyCollide, 1))
+        >>> entity2
+        KeyCollide(_key=Key('KeyCollide', 1), key='Go slow')
+
+    .. automethod:: _get_kind
+
+    Args:
+        key (Key): Datastore key for this entity (kind must match this model).
+            If ``key`` is used, ``id`` and ``parent`` must be unset or
+            :data:`None`.
+        id (str): Key ID for this model. If ``id`` is used, ``key`` must be
+            :data:`None`.
+        parent (Key): The parent model or :data:`None` for a top-level model.
+            If ``parent`` is used, ``key`` must be :data:`None`.
+        namespace (str): Namespace for the entity key.
+        app (str): Application ID for the entity key.
+        kwargs (Dict[str, Any]): Additional keyword arguments. These should map
+            to properties of this model.
+
+    Raises:
+        .BadArgumentError: If the constructor is called with ``key`` and one
+            of ``id``, ``app``, ``namespace`` or ``parent`` specified.
```

Comment:
Mention TypeError if kwarg isn't a property, or no?

Reply:
It'd be an `AttributeError` I think. Did you have a code snippet in mind?

Reply:
Here is what happens against the original implementation:

```python
from google.appengine.ext import ndb


class A(ndb.Model):
    b = ndb.IntegerProperty()


print(A)
print(A(b=11))
print(A(c=12))
```

yields

```
A<b=IntegerProperty('b')>
A(b=11)
Traceback (most recent call last):
  ...
  File ".../google/appengine/ext/ndb/model.py", line 2947, in __init__
    self._set_attributes(kwds)
  File ".../google/appengine/ext/ndb/model.py", line 2990, in _set_attributes
    prop = getattr(cls, name)  # Raises AttributeError for unknown properties.
AttributeError: type object 'A' has no attribute 'c'
```

Reply:
I was just referencing this:

https://github.com/googleapis/google-cloud-python/blob/639234a9294375158522c23c3795c976ceb78b75/ndb/src/google/cloud/ndb/model.py#L3254

Which throws TypeError. AttributeError is probably better, though.

Reply:
Gotcha, that's this instead:

```python
from google.appengine.ext import ndb


class A(ndb.Model):
    b = ndb.IntegerProperty()
    c = 98909


print(A)
print(A(b=11))
print(A(c=12))
```

yields

```
A<b=IntegerProperty('b')>
A(b=11)
Traceback (most recent call last):
  ...
  File ".../google/appengine/ext/ndb/model.py", line 2947, in __init__
    self._set_attributes(kwds)
  File ".../google/appengine/ext/ndb/model.py", line 2992, in _set_attributes
    raise TypeError('Cannot set non-property %s' % name)
TypeError: Cannot set non-property c
```

(This is in a unit test, but wanted to add it for completeness.)

I can add doctests to the constructor to avoid staleness?

Reply:
Ah ok.

Reply:
Added doctest.